### PR TITLE
Autogenerated types for Snowbound API

### PR DESCRIPTION
**Pull request by floatplane**
This PR does the following:

1. Adds scripts to take the Snowbound OpenAPI definition and turn it into type definitions and wrapper classes. The scripts manually insert `token` params into each API path because the package that we're using to generate code (`openapi-typescript-codegen`) doesn't fully understand the OpenAPI spec*. In other words, I don't think Snowbound was actually doing anything wrong.
1. Fixes various bits of infra (eslint, prettier) to not touch the generated code

Everything inside `types/generated` is in fact generated, so if you have comments on those files I will probably not act upon them :) 

--- 

*the part of the spec that the codegen package doesn't seem to pick up on is this:
 ```
        "securitySchemes": {
            "OAuth2PasswordBearerWithCookie": {
                "type": "oauth2",
                "flows": {
                    "password": {
                        "scopes": {},
                        "tokenUrl": "/token"
                    }
                }
            },
            "APIKeyQuery": {
                "type": "apiKey",
                "in": "query",
                "name": "token"
            }
        }
 ```
which I believe should be interpreted as "add a `token` query param as optional to every request method"

### CODE REVIEW

File: scripts/gen-openapi.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Download Snowbounds published API and fix it up locally
```
Comment (stevekuznetsov) [inline scripts/gen-openapi.sh:1]: Could we please:


```sh
#!/bin/bash

set -o errexit
set -o nounset
set -o pipefail
```

File: scripts/getSnowboundOpenAPISpec [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+npx ts-node 'scripts/getSnowboundOpenAPISpec.ts'
```
Comment (stevekuznetsov) [inline scripts/getSnowboundOpenAPISpec:2]: Why is this a separate shell script? Could we just inline this?

File: scripts/gen-openapi.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+tmpfile=$(mktemp /tmp/openapi.XXXXXX)
```
Comment (stevekuznetsov) [inline scripts/gen-openapi.sh:2]:
```sh
tmpfile="$(mktemp /tmp/openapi.XXXXXX)"

function cleanup() {
    rm -rf "${tmpfile}"
}

trap EXIT cleanup
```

File: scripts/gen-openapi.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+./scripts/getSnowboundOpenAPISpec > $tmpfile
+
+# Generate TS type definitions from Snowbound's published API
+yarn openapi \
+    --input $tmpfile \
```
Comment (stevekuznetsov) [inline scripts/gen-openapi.sh:7]:
```suggestion
    --input "${tmpfile}" \
```

File: scripts/gen-openapi.sh (hunk 3 above)
Comment (stevekuznetsov) [inline scripts/gen-openapi.sh:2]:
```suggestion
tmpfile="$(mktemp /tmp/openapi.XXXXXX)"
```

File: scripts/gen-openapi.sh [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# Download Snowbounds published API and fix it up locally
+tmpfile=$(mktemp /tmp/openapi.XXXXXX)
+./scripts/getSnowboundOpenAPISpec > $tmpfile
```
Comment (stevekuznetsov) [inline scripts/gen-openapi.sh:3]:
```suggestion
npx ts-node 'scripts/getSnowboundOpenAPISpec.ts' > "${tmpfile}"
```

File: types/generated/snowbound/core/ApiError.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/* istanbul ignore file */
```
Comment (stevekuznetsov) [inline types/generated/snowbound/core/ApiError.ts:1]: super-nit: not sure we _need_ another level of directory nesting for `generated` - are we going to co-locate generated and non-generated types for one API?

File: scripts/getSnowboundOpenAPISpec (hunk 2 above)
Comment (floatplane) [reply scripts/getSnowboundOpenAPISpec:2]: 🤷  sure

File: scripts/gen-openapi.sh (hunk 3 above)
Comment (floatplane) [reply scripts/gen-openapi.sh:2]: are you really worried that `mktemp` will produce a filename with a space in it?

File: types/generated/snowbound/core/ApiError.ts (hunk 6 above)
Comment (floatplane) [reply types/generated/snowbound/core/ApiError.ts:1]: I don't know, maybe? I think it's useful to have a clear delineation between "things you should edit" and "things you shouldn't".

File: scripts/gen-openapi.sh (hunk 3 above)
Comment (stevekuznetsov) [reply scripts/gen-openapi.sh]: Rule number one of bash - never trust anything :) In this context, you never have to wonder - why is it valid for _this_ variable to be unquoted? Simply not a useful distinction

File: types/generated/snowbound/core/ApiError.ts (hunk 6 above)
Comment (stevekuznetsov) [reply types/generated/snowbound/core/ApiError.ts:1]: I see. Our GraphQL generated stuff is in the wrong place then - I am used to CI that will re-gen and ensure that things are up-to-date, as well. In order for CI to not trigger diffs, I'd check in the OpenAPI schema (like we did for GraphQL) and validate that the schema -> code generation is correct at merge-time.

**Pull request merged**